File: .claude/skills/smm/scripts/plan_manager.py
#!/usr/bin/env python3
"""
plan_manager.py — Gestisce il file editorial-plan.md
Uso:
  python plan_manager.py --check              Verifica stato piano
  python plan_manager.py --today              Mostra post di oggi
  python plan_manager.py --status DATE PLATFORM STATUS  Aggiorna status
  python plan_manager.py --metrics-check      Giorni da ultima revisione
"""

import argparse
import re
import sys
from datetime import date, datetime
from pathlib import Path

PLAN_PATH = Path(__file__).parents[4] / "plans" / "editorial-plan.md"

STATUS_SYMBOLS = {"scheduled": "⬜", "created": "🟡", "published": "✅"}


def _read_raw():
    if not PLAN_PATH.exists():
        print("❌ Piano non trovato. Lancia /smm per crearlo.", file=sys.stderr)
        sys.exit(1)
    return PLAN_PATH.read_text(encoding="utf-8")


def _parse_frontmatter(raw):
    match = re.match(r"^---\n(.*?)\n---", raw, re.DOTALL)
    if not match:
        return {}
    fm = {}
    for line in match.group(1).splitlines():
        if ":" in line:
            key, _, val = line.partition(":")
            fm[key.strip()] = val.strip()
    return fm


def _parse_table_rows(raw):
    rows = []
    for line in raw.splitlines():
        if line.startswith("|") and not line.startswith("| Date") and "---" not in line:
            parts = [p.strip() for p in line.strip("|").split("|")]
            if len(parts) >= 8:
                rows.append({
                    "date": parts[0],
                    "platform": parts[1],
                    "stage": parts[2],
                    "topic": parts[3],
                    "format": parts[4],
                    "hook": parts[5],
                    "cta": parts[6],
                    "status": parts[7],
                })
    return rows


def check():
    raw = _read_raw()
    fm = _parse_frontmatter(raw)
    rows = _parse_table_rows(raw)
    print(f"✓ Piano trovato: {fm.get('period', 'N/A')}")
    print(f"  Obiettivo: {fm.get('monthly_goal', 'N/A')}")
    print(f"  Post totali: {len(rows)}")
    completed = sum(1 for r in rows if r["status"] == STATUS_SYMBOLS["published"])
    created = sum(1 for r in rows if r["status"] == STATUS_SYMBOLS["created"])
    scheduled = sum(1 for r in rows if r["status"] == STATUS_SYMBOLS["scheduled"])
    print(f"  ✅ Pubblicati: {completed}  🟡 Creati: {created}  ⬜ Da fare: {scheduled}")


def today_posts():
    raw = _read_raw()
    rows = _parse_table_rows(raw)
    today_str = date.today().isoformat()
    results = [r for r in rows if r["date"] == today_str and r["status"] == STATUS_SYMBOLS["scheduled"]]
    if not results:
        print(f"Nessun post schedulato per oggi ({today_str}).")
    else:
        print(f"Post di oggi ({today_str}):")
        for r in results:
            print(f"  • {r['platform']} ({r['stage']}) — {r['topic']} [{r['format']}]")
    return results


def update_status(target_date, platform, new_status):
    raw = _read_raw()
    symbol = STATUS_SYMBOLS.get(new_status)
    if not symbol:
        print(f"❌ Status non valido: {new_status}. Usa: scheduled, created, published", file=sys.stderr)
        sys.exit(1)

    lines = raw.splitlines()
    updated = False
    for i, line in enumerate(lines):
        if line.startswith("|") and target_date in line and platform in line:
            parts = line.split("|")
            if len(parts) >= 9:
                parts[-2] = f" {symbol} "
                lines[i] = "|".join(parts)
                updated = True
                break

    if not updated:
        print(f"⚠️  Riga non trovata per {target_date} / {platform}", file=sys.stderr)
        sys.exit(1)

    PLAN_PATH.write_text("\n".join(lines), encoding="utf-8")
    print(f"✓ Status aggiornato: {target_date} {platform} → {symbol}")


def metrics_check():
    raw = _read_raw()
    fm = _parse_frontmatter(raw)
    last_review = fm.get("last_metrics_review", "")
    if not last_review:
        print("0")
        return
    try:
        last_dt = datetime.fromisoformat(last_review).date()
        days = (date.today() - last_dt).days
        print(str(days))
    except ValueError:
        print("0")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Plan Manager")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--today", action="store_true")
    parser.add_argument("--status", nargs=3, metavar=("DATE", "PLATFORM", "STATUS"))
    parser.add_argument("--metrics-check", action="store_true")
    args = parser.parse_args()

    if args.check:
        check()
    elif args.today:
        today_posts()
    elif args.status:
        update_status(*args.status)
    elif args.metrics_check:
        metrics_check()
    else:
        parser.print_help()

File: .claude/skills/smm/scripts/plan_manager.js
#!/usr/bin/env node
/**
 * plan_manager.js — Gestisce il file editorial-plan.md
 * Uso:
 *   node plan_manager.js --check
 *   node plan_manager.js --today
 *   node plan_manager.js --status DATE PLATFORM STATUS
 *   node plan_manager.js --metrics-check
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PLAN_PATH = resolve(__dirname, '../../../../plans/editorial-plan.md');

const STATUS_SYMBOLS = { scheduled: '⬜', created: '🟡', published: '✅' };

function readRaw() {
  if (!existsSync(PLAN_PATH)) {
    console.error('❌ Piano non trovato. Lancia /smm per crearlo.');
    process.exit(1);
  }
  return readFileSync(PLAN_PATH, 'utf8');
}

function parseFrontmatter(raw) {
  const match = raw.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return {};
  return Object.fromEntries(
    match[1].split('\n').filter(l => l.includes(':')).map(l => {
      const [k, ...v] = l.split(':');
      return [k.trim(), v.join(':').trim()];
    })
  );
}

function parseTableRows(raw) {
  return raw.split('\n')
    .filter(l => l.startsWith('|') && !l.startsWith('| Date') && !l.includes('---'))
    .map(l => l.split('|').map(p => p.trim()).filter(Boolean))
    .filter(parts => parts.length >= 8)
    .map(parts => ({
      date: parts[0], platform: parts[1], stage: parts[2], topic: parts[3],
      format: parts[4], hook: parts[5], cta: parts[6], status: parts[7],
    }));
}

function check() {
  const raw = readRaw();
  const fm = parseFrontmatter(raw);
  const rows = parseTableRows(raw);
  console.log(`✓ Piano trovato: ${fm.period ?? 'N/A'}`);
  console.log(`  Obiettivo: ${fm.monthly_goal ?? 'N/A'}`);
  console.log(`  Post totali: ${rows.length}`);
  const published = rows.filter(r => r.status === STATUS_SYMBOLS.published).length;
  const created = rows.filter(r => r.status === STATUS_SYMBOLS.created).length;
  const scheduled = rows.filter(r => r.status === STATUS_SYMBOLS.scheduled).length;
  console.log(`  ✅ Pubblicati: ${published}  🟡 Creati: ${created}  ⬜ Da fare: ${scheduled}`);
}

function todayPosts() {
  const raw = readRaw();
  const rows = parseTableRows(raw);
  const todayStr = new Date().toISOString().split('T')[0];
  const results = rows.filter(r => r.date === todayStr && r.status === STATUS_SYMBOLS.scheduled);
  if (!results.length) {
    console.log(`Nessun post schedulato per oggi (${todayStr}).`);
  } else {
    console.log(`Post di oggi (${todayStr}):`);
    results.forEach(r => console.log(`  • ${r.platform} (${r.stage}) — ${r.topic} [${r.format}]`));
  }
  return results;
}

function updateStatus(targetDate, platform, newStatus) {
  const raw = readRaw();
  const symbol = STATUS_SYMBOLS[newStatus];
  if (!symbol) {
    console.error(`❌ Status non valido: ${newStatus}. Usa: scheduled, created, published`);
    process.exit(1);
  }
  const lines = raw.split('\n');
  let updated = false;
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].startsWith('|') && lines[i].includes(targetDate) && lines[i].includes(platform)) {
      const parts = lines[i].split('|');
      if (parts.length >= 9) {
        parts[parts.length - 2] = ` ${symbol} `;
        lines[i] = parts.join('|');
        updated = true;
        break;
      }
    }
  }
  if (!updated) {
    console.error(`⚠️  Riga non trovata per ${targetDate} / ${platform}`);
    process.exit(1);
  }
  writeFileSync(PLAN_PATH, lines.join('\n'), 'utf8');
  console.log(`✓ Status aggiornato: ${targetDate} ${platform} → ${symbol}`);
}

function metricsCheck() {
  const raw = readRaw();
  const fm = parseFrontmatter(raw);
  const lastReview = fm.last_metrics_review ?? '';
  if (!lastReview) { console.log('0'); return; }
  try {
    const lastDt = new Date(lastReview);
    const days = Math.floor((Date.now() - lastDt.getTime()) / 86400000);
    console.log(String(days));
  } catch {
    console.log('0');
  }
}

// CLI
const args = process.argv.slice(2);
const get = (flag) => { const i = args.indexOf(flag); return i !== -1 ? args[i + 1] : null; };

if (args.includes('--check')) check();
else if (args.includes('--today')) todayPosts();
else if (args.includes('--status')) {
  const i = args.indexOf('--status');
  updateStatus(args[i + 1], args[i + 2], args[i + 3]);
} else if (args.includes('--metrics-check')) metricsCheck();
else console.log('Uso: node plan_manager.js --check | --today | --status DATE PLATFORM STATUS | --metrics-check');
